macros/src/lib.rs

```diff
@@ -4,513 +4,560 @@ mod assembly;
+            accessors.push(quote! {
+                pub fn #get_name(&self) -> #ty {
+                    self.#ident.clone()
+                }
+            });
+
+            if region.single_block {
+                let get_name = format_ident!("get_{}", ident);
+                accessors.push(quote! {
+                    pub fn #get_name(&self) -> tir_core::BlockRef {
+                        self.#ident.first().unwrap()
+                    }
+                });
+            }
+        }
     }
 
     quote! {
-        #(#impls)*
+        #(#accessors)*
     }
 }
 
-fn build_operand_accessors(operands: &[OperandField]) -> proc_macro2::TokenStream {
-    let mut impls = vec![];
+fn build_attr_accessors(attrs: &[Attr]) -> proc_macro2::TokenStream {
+    let mut attr_accessors = vec![];
 
-    for (id, operand) in operands.iter().enumerate() {
-        let get_func = format_ident!("get_{}", operand.ident);
-        // TODO implement setters!
-        let set_func = format_ident!("set_{}", operand.ident);
-        let funcs = quote! {
-            pub fn #get_func(&self) -> &Operand {
-                &self.operation.operands[#id]
-            }
-            pub fn #set_func(&mut self) {
-                todo!()
+    for attr in attrs {
+        let getter_name = format_ident!("get_{}_attr", attr.0);
+        let setter_name = format_ident!("set_{}_attr", attr.0);
+        let attr_str = attr.0.to_string();
+
+        attr_accessors.push(quote! {
+            pub fn #getter_name<'a>(&'a self) -> &'a tir_core::Attr {
+                self.r#impl.attrs.get(#attr_str).unwrap()
             }
-        };
 
-        impls.push(funcs);
+            pub fn #setter_name<T>(&mut self, value: T) where tir_core::Attr: From<T> {
+                let attr = tir_core::Attr::from(value);
+                self.r#impl.attrs.insert(#attr_str.to_string(), attr);
+            }
+        });
     }
 
     quote! {
-        #(#impls)*
+        #(#attr_accessors)*
     }
 }
 
-fn build_region_accessors(regions: &[RegionField]) -> proc_macro2::TokenStream {
-    let mut impls = vec![];
-
-    for (id, region) in regions.iter().enumerate() {
-        let get_func = format_ident!("get_{}_region", region.ident);
-        let func = quote! {
-            pub fn #get_func(&self) -> RegionRef {
-                self.operation
-                    .regions[#id]
-                    .clone()
-            }
-        };
-        impls.push(func);
-
-        if region.single_block {
-            let get_func = format_ident!("get_{}", region.ident);
-            let func = quote! {
-                pub fn #get_func(&self) -> BlockRef {
-                    self.operation
-                        .regions[#id]
-                        .borrow()
-                        .get_blocks()
-                        .first()
-                        .unwrap()
-                        .clone()
+fn build_return_type_accessor(fields: &[OpFieldReceiver]) -> proc_macro2::TokenStream {
```

Comment:
Making `return_type` a field attribute allows for multiple return types, is it intentional?

Reply:
For example, this code compiles:
```rust
#[derive(Op, Assembly)]
#[operation(name = "const", known_attrs(value: IntegerAttr))]
pub struct ConstOp {
    #[ret_type]
    return_type: Type,
    #[ret_type]
    rety: Type,
    r#impl: OpImpl,
}
```

Reply:
Maybe we should make it function-level attribute instead?

Reply:
maybe https://github.com/perf-toolbox/tir/pull/37/files#diff-4feb5862b96e70d6a225028210762f9a2d19ea5e35860fe78d211fb607c82b26R182-R192 ?

Reply:
> Maybe we should make it function-level attribute instead?

Not possible for derive macros